modules/sitemaps/sitemaps.php

```diff
@@ -0,0 +1,136 @@
+<?php
+/**
+ * @package Polylang
+ */
+
+/**
+ * Handles the core sitemaps.
+ *
+ * @since 2.8
+ */
+class PLL_Sitemaps {
+	/**
+	 * A reference to teh PLL_Model instance.
+	 *
+	 * @since 2.8
+	 *
+	 * @var PLL_Model
+	 */
+	protected $model;
+
+	/**
+	 * A reference to the current language.
+	 *
+	 * @since 2.8
+	 *
+	 * @var PLL_Language
+	 */
+	protected $curlang;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 2.8
+	 *
+	 * @param object $polylang Main Polylang object.
+	 */
+	public function __construct( $polylang ) {
+		$this->model = &$polylang->model;
+		$this->curlang = &$polylang->curlang;
+	}
+
+	/**
+	 * Setups actions and filters.
+	 *
+	 * @since 2.8
+	 */
+	public function init() {
+		add_action( 'template_redirect', array( $this, 'set_language' ), 5 ); // Before WP_Sitemaps.
+		add_filter( 'pll_home_url_white_list', array( $this, 'home_url_white_list' ) );
+		add_filter( 'rewrite_rules_array', array( $this, 'rewrite_rules' ) );
+		add_filter( 'wp_sitemaps_register_providers', array( $this, 'providers' ), 99 ); // 99 in an attempt to have all sitemaps providers in our filter.
+	}
+
+	/**
+	 * Sets the language for the current sitemap being rendered.
+	 *
+	 * @since 2.8
+	 */
+	public function set_language() {
+		$qv = get_query_var( 'sitemap' );
+		if ( $qv ) {
+			$arr = explode( '-', $qv );
+			$lang = end( $arr );
+			$this->curlang = $this->model->get_language( $lang );
+		}
+	}
+
+	/**
+	 * Whitelists the home url filter for the sitemaps
```

Comment:
Do you mean that it adds the wp-sitemap calls to the `home_url()` WordPress function in our whitelist for results that should not be filtered by language?

Reply:
I don't understand the issue you tried to explain during the DM and what it is the goal to whitelist `sitemaps` filenames.

Reply:
By default, our `home_url` filter is active for themes and some whitelisted functions of WP core, but not for plugins. For sitemaps, we always need to filter the home url. So I am whitelisting all files with 'sitemaps' somewhere in their path. I don't believe it's risky. If some conflict arises at some time we could try to narrow the whitelist.

Reply:
Okay, so the whitelisted files and functions will always be filtered by language, is that right?

Reply:
Any call to `home_url()` done from the whitelisted files and functions are indeed filtered.